### PR TITLE
Central function to retrieve item or thread data

### PR DESCRIPTION
**Pull request by annando**
Additionally this function has been implemented in the API.

### CODE REVIEW

**Comment by annando** [comment]
@Rudloff I guess that the tests are failing because I changed many things in the item query. Could you help to fix this?

File: include/api.php [hunk 1]
```diff
@@ -1689,33 +1678,24 @@ function api_statuses_home_timeline($type)
...
-		$sql_extra .= ' AND `item`.`parent` = `item`.`id`';
+		$condition[0] .= ' AND `item`.`parent` = `item`.`id`';
 	}
 	if ($conversation_id > 0) {
-		$sql_extra .= ' AND `item`.`parent` = ' . intval($conversation_id);
+		$condition[0] .= " AND `item`.`parent` = ?";
+		$condition[] = $conversation_id;
 	}
 
-	$r = q("SELECT `item`.* FROM `item`
```
Comment (Rudloff) [inline include/api.php]: You removed the `$r` variable here but the foreach loop under still uses it.

Comment (annando) [reply include/api.php]: Where have you seen this loop? Could you point me to the line?

Comment (Rudloff) [reply include/api.php]: Here: https://github.com/annando/friendica/blob/50dfc7d36ef08d1be8c50b0612ca0d8613c2aab5/include/api.php#L1702

Comment (annando) [reply include/api.php]: And fixed.

File: include/api.php [hunk 2]
```diff
@@ -1975,24 +1909,22 @@ function api_statuses_show($type)
...
+	} else {
+		$condition = ['id' => $id, 'verb' => ACTIVITY_POST];
+		$params = [];
+	}
+
+	$statuses = Item::select(api_user(), [], $condition, $params);
 
 	/// @TODO How about copying this to above methods which don't check $r ?
-	if (!DBM::is_result($r)) {
+	if (!DBM::is_result($items)) {
```
Comment (Rudloff) [inline include/api.php]: Shouldn't you use `$statuses` here?

File: include/api.php [hunk 3]
```diff
@@ -2057,33 +1989,22 @@ function api_conversation_show($type)
...
+		$condition[0] .= " AND `item`.`id` <= ?";
+		$condition[] = $max_id;
+	}
 
-	if (!DBM::is_result($r)) {
-		throw new BadRequestException("There is no status with this id.");
+	$params = ['order' => ['id' => true], 'limit' => [$start, $count]];
+	$statuses = Item::select(api_user(), [], $condition, $params);
+
+	if (!DBM::is_result($items)) {
```
Comment (Rudloff) [inline include/api.php]: Same problem here, it should be `$statuses` I think.

**Comment by Rudloff** [comment]
I think `api_favorites_create_destroy()` tests fail because item 1 has buggy user IDs.
A quick way to fix this is to use another item from the dataset:
```diff
diff --git a/tests/ApiTest.php b/tests/ApiTest.php
index f0e27b4af..c21f651d5 100644
--- a/tests/ApiTest.php
+++ b/tests/ApiTest.php
@@ -1871,7 +1871,7 @@ class ApiTest extends DatabaseTest
                $this->app->argv[1] = '1.1';
                $this->app->argv[3] = 'create';
                $this->app->argc = 10;
-               $_REQUEST['id'] = 1;
+               $_REQUEST['id'] = 3;
                $result = api_favorites_create_destroy('json');
                $this->assertStatus($result['status']);
        }
@@ -1885,7 +1885,7 @@ class ApiTest extends DatabaseTest
                $this->app->argv[1] = '1.1';
                $this->app->argv[3] = 'create';
                $this->app->argc = 10;
-               $_REQUEST['id'] = 1;
+               $_REQUEST['id'] = 3;
                $result = api_favorites_create_destroy('rss');
                $this->assertXml($result, 'status');
        }
@@ -1899,7 +1899,7 @@ class ApiTest extends DatabaseTest
                $this->app->argv[1] = '1.1';
                $this->app->argv[3] = 'destroy';
                $this->app->argc = 10;
-               $_REQUEST['id'] = 1;
+               $_REQUEST['id'] = 3;
                $result = api_favorites_create_destroy('json');
                $this->assertStatus($result['status']);
        }
```
(Until someone someone who understands how these IDs work can update the dataset.)

**Comment by annando** [comment]
I do understand the whole data model - but I cannot easily test the test stuff.

**Comment by annando** [comment]
It's getting better.

**Comment by Rudloff** [comment]
This should fix the `api_statuses_public_timeline()` and `api_statuses_networkpublic_timeline()` tests : 
```diff
diff --git a/tests/datasets/api.yml b/tests/datasets/api.yml
index 9cacab714..9ba5ec387 100644
--- a/tests/datasets/api.yml
+++ b/tests/datasets/api.yml
@@ -48,7 +48,7 @@ contact:
         network: dfrn
     -
         id: 44
-        uid: 42
+        uid: 0
         name: Friend contact
         nick: friendcontact
         self: false
@@ -74,6 +74,7 @@ item:
         author-link: http://localhost/profile/selfcontact
         wall: true
         starred: true
+        origin: true
         allow_cid: ''
         allow_gid: ''
         deny_cid: ''
@@ -92,6 +93,7 @@ item:
         author-link: http://localhost/profile/selfcontact
         wall: true
         starred: false
+        origin: true
     -
         id: 3
         visible: true
@@ -106,20 +108,22 @@ item:
         author-link: http://localhost/profile/othercontact
         wall: true
         starred: false
+        origin: true
     -
         id: 4
         visible: true
-        contact-id: 43
-        author-id: 43
+        contact-id: 44
+        author-id: 44
         owner-id: 42
         uid: 42
         verb: http://activitystrea.ms/schema/1.0/post
         unseen: false
-        body: Other user reply
+        body: Friend user reply
         parent: 1
         author-link: http://localhost/profile/othercontact
         wall: true
         starred: false
+        origin: true
     -
         id: 5
         visible: true
@@ -134,6 +138,7 @@ item:
         author-link: http://localhost/profile/othercontact
         wall: true
         starred: false
+        origin: true
         allow_cid: ''
         allow_gid: ''
         deny_cid: ''
@@ -152,24 +157,31 @@ item:
         author-link: http://localhost/profile/othercontact
         wall: true
         starred: false
+        origin: true
 
 thread:
     -
         iid: 1
         visible: true
         contact-id: 42
+        author-id: 42
+        owner-id: 42
         uid: 42
         wall: true
     -
         iid: 3
         visible: true
         contact-id: 43
+        author-id: 43
+        owner-id: 43
         uid: 0
         wall: true
     -
         iid: 6
         visible: true
         contact-id: 44
+        author-id: 44
+        owner-id: 44
         uid: 0
         wall: true
```
(There were some missing fields and `testApiStatusesPublicTimeline()` needs a reply to item 1 from a contact with rel 2 or 3.)

**Comment by annando** [comment]
I'm now done with this PR. The next changes will be covered in following PRs.

**Pull request merged**